Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/Migration/Tools/MigrationTool.cs

```diff
@@ -0,0 +1,264 @@
+            {
+                EditorSceneManager.OpenScene(Directory.GetCurrentDirectory() + "/" + previousScenePath);
+            }
+            return true;
+        }
+
+        private void AddAllAssetsOfTypeForMigration(Type[] types)
+        {
+            if (types == null)
+            {
+                return;
+            }
+
+            var assetPaths = FindAllAssetsOfType(types);
+            if (assetPaths != null)
+            {
+                foreach (var path in assetPaths)
+                {
+                    TryAddObjectForMigration(AssetDatabase.LoadMainAssetAtPath(path));
+                }
+            }
+        }
+
+        private bool SetMigrationHandlerInstance(Type type)
+        {
+            if (type == null || !typeof(IMigrationHandler).IsAssignableFrom(type))
+            {
+                Debug.LogError($"{type.Name} is not a valid implementation of IMigrationHandler");
+                return false;
+            }
+
+            if (!migrationHandlerTypes.Contains(type))
+            {
+                Debug.LogError($"{type.Name} might not be a valid implementation of IMigrationHandler");
+                return false;
+            }
+
+            try
+            {
+                migrationHandlerInstance = Activator.CreateInstance(type) as IMigrationHandler;
+            }
+            catch (Exception)
+            {
+                Debug.LogError("Selected MigrationHandler implementation could not be instanciated");
+                return false;
+            }
+            return true;
+        }
+
+        private void RefreshAvailableTypes()
+        {
+            var type = typeof(IMigrationHandler);
+            migrationHandlerTypes = AppDomain.CurrentDomain.GetAssemblies()
+                .SelectMany(x => x.GetLoadableTypes())
+                .Where(x => type.IsAssignableFrom(x) && !x.IsInterface && !x.IsAbstract).ToList();
+        }
+
+        private void MigrateScene(String path)
+        {
+            if (!AssetDatabase.LoadAssetAtPath(path, typeof(SceneAsset)))
+            {
+                return;
+            }
+            Scene scene = EditorSceneManager.OpenScene(path);
+
+            foreach (var parent in scene.GetRootGameObjects())
+            {
+                MigrateGameObjectHierarchy(parent);
+            }
+            EditorSceneManager.SaveScene(scene);
+        }
+
+        private void MigratePrefab(String path)
+        {
+            if (!AssetDatabase.LoadAssetAtPath(path, typeof(GameObject)))
+            {
+                return;
+            }
+            var parent = UnityEditor.PrefabUtility.LoadPrefabContents(path);
+
+            MigrateGameObjectHierarchy(parent);
+
+            UnityEditor.PrefabUtility.SaveAsPrefabAsset(parent, path);
+            PrefabUtility.UnloadPrefabContents(parent);
+        }
+
+        private void MigrateGameObjectHierarchy(GameObject parent)
+        {
+            if (!parent)
```

Comment:
More general meta comment that I'm leaving here without having fully traced all of the call/paths that invoke MigrateGameObjectHierarchy or the other functions here. A lot of these seem to do null checks or validity checks at the function level.

This is generally okay when you're getting data from 'untrusted' sources (i.e. if these are meant to be invoked by external code or through an API that we expose). However if this is an internal function it's usually better to trust that your other stuff doesn't call you incorrectly (i.e. it's non-null). Otherwise what tends to happen is that  you get a lot of extra code at the beginning of every function that's checking for null/validity, which builds up as call depth increases/other things repeatedly do the same check.

More of a comment for going forward (and also a question for if this is really intended to be used by non-internal code).

Reply:
Yeah, I was unsure about the approach. also prefere to avoid the extra check.